feat: add vue/no-undef-directives rule (#2990)
vuejs/eslint-plugin-vue: docs/rules/no-undef-directives.md

## :book: Rule Details

This rule reports directives that are used in the `<template>`, but that are not registered in the `<script setup>` or the Options API's `directives` section.

Undefined directives will be resolved from globally registered directives. However, if you are not using global directives, you can use this rule to prevent run-time errors.

<eslint-code-block :rules="{'vue/no-undef-directives': ['error']}">

```vue
<script setup>
import vFocus from './vFocus';
</script>

<template>
  <!-- ✓ GOOD -->
  <input v-focus>

  <!-- ✗ BAD -->
  <div v-foo></div>
</template>
```

</eslint-code-block>

<eslint-code-block :rules="{'vue/no-undef-directives': ['error']}">

```vue
<template>
  <!-- ✓ GOOD -->
  <input v-focus>

  <!-- ✗ BAD -->
  <div v-foo></div>
</template>

<script>
import vFocus from './vFocus';

export default {
  directives: {
    focus: vFocus
  }
}
</script>
```

</eslint-code-block>

## :wrench: Options

```json
{
  "vue/no-undef-directives": ["error", {
    "ignorePatterns": ["foo"]
  }]
}
```

- `ignorePatterns` (`string[]`) ... An array of regex pattern strings to ignore.

> ### `ignorePatterns`

Reviewer: ```suggestion
### "ignorePatterns": ["foo"]
```

> ### "ignorePatterns": ["foo"]

Reviewer: Sorry, I forgot the code quotes in my earlier suggestion

```suggestion
### `"ignorePatterns": ["foo"]`
```

- :exclamation: <badge text="This rule has not been released yet." vertical="middle" type="error"> _**This rule has not been released yet.**_ </badge>

## :book: Rule Details

This rule reports directives that are used in the `<template>`, but that are not registered in the `<script setup>` or the Options API's `directives` section.

Undefined directives will be resolved from globally registered directives. However, if you are not using global directives, you can use this rule to prevent runtime errors.

<eslint-code-block :rules="{'vue/no-undef-directives': ['error']}">

```vue
<script setup>
import vFocus from './vFocus';
</script>

<template>
  <!-- ✓ GOOD -->
  <input v-focus>

  <!-- ✗ BAD -->
  <div v-foo></div>
</template>
```

</eslint-code-block>

<eslint-code-block :rules="{'vue/no-undef-directives': ['error']}">

```vue
<template>
  <!-- ✓ GOOD -->
  <input v-focus>

  <!-- ✗ BAD -->
  <div v-foo></div>
</template>

<script>
import vFocus from './vFocus';

export default {
  directives: {
    focus: vFocus
  }
}
</script>
```

</eslint-code-block>

## :wrench: Options

```json
{
  "vue/no-undef-directives": ["error", {
    "ignore": ["foo"]
  }]
}
```

> - `"ignore"` (`string[]`) An array of tags or regular expression patterns (e.g. `"/^custom-/"`) that ignore these rules. This option will check both kebab-case and PascalCase versions of the given tag names. Default is empty.

Reviewer: ```suggestion
- `"ignore"` (`string[]`) An array of directive names or regular expression patterns (e.g. `"/^custom-/"`) that ignore these rules. This option will check both kebab-case and PascalCase versions of the given directive names. Default is empty.
```
